authentication/oauth.go

```diff
@@ -193,21 +204,86 @@ func (o *OAuth) RevokeRefreshToken(ctx context.Context, body oauth.RevokeRefresh
 }
 
 func (o *OAuth) addClientAuthentication(params oauth.ClientAuthentication, body url.Values, required bool) error {
-	if params.ClientID != "" {
-		body.Set("client_id", params.ClientID)
-	} else {
-		body.Set("client_id", o.authentication.clientID)
+	clientID := params.ClientID
+	if params.ClientID == "" {
+		clientID = o.authentication.clientID
 	}
+	body.Set("client_id", clientID)
 
-	if params.ClientSecret != "" && body.Get("client_secret") == "" {
-		body.Set("client_secret", params.ClientSecret)
-	} else if o.authentication.clientSecret != "" {
-		body.Set("client_secret", o.authentication.clientSecret)
+	clientSecret := params.ClientSecret
+	if params.ClientSecret == "" {
+		clientSecret = o.authentication.clientSecret
 	}
 
-	if required && body.Get("client_secret") == "" {
-		return errors.New("client_secret is required but not provided")
+	switch {
+	case o.authentication.clientAssertionSigningKey != "":
+		clientAssertion, err := createClientAssertion(
+			o.authentication.clientAssertionSigningAlg,
+			o.authentication.clientAssertionSigningKey,
+			clientID,
+			o.authentication.domain,
+		)
+		if err != nil {
+			return err
+		}
+
+		body.Set("client_assertion", clientAssertion)
+		body.Set("client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer")
+		break
+	case params.ClientAssertion != "":
+		body.Set("client_assertion", params.ClientAssertion)
+		body.Set("client_assertion_type", params.ClientAssertionType)
+		break
+	case clientSecret != "":
+		body.Set("client_secret", clientSecret)
+		break
+	}
+
+	if required && (body.Get("client_secret") == "" && body.Get("client_assertion") == "") {
+		return errors.New("client_secret or client_assertion is required but not provided")
 	}
 
 	return nil
 }
+
+func determineAlg(alg string) (jwa.SignatureAlgorithm, error) {
+	switch alg {
+	case "HS256":
+		return jwa.HS256, nil
+	case "RS256":
+		return jwa.RS256, nil
+	default:
+		return "", fmt.Errorf("Unsupported algorithm %s provided", alg)
+	}
+}
+
+func createClientAssertion(clientAssertionSigningAlg, clientAssertionSigningKey, clientID, domain string) (string, error) {
+	alg, err := determineAlg(clientAssertionSigningAlg)
+	if err != nil {
+		return "", err
+	}
+
+	key, err := jwk.ParseKey([]byte(clientAssertionSigningKey), jwk.WithPEM(true))
+	if err != nil {
+		return "", err
+	}
+
+	token, err := jwt.NewBuilder().
+		IssuedAt(time.Now()).
+		Subject(clientID).
+		JwtID(uuid.New().String()).
+		Issuer(clientID).
+		Audience([]string{domain}).
+		Expiration(time.Now().Add(2 * time.Minute)).
```

Comment:
Just out of curiosity, where does this 2 minute exp come from?

Reply:
I was aligning with the value set by the node-auth0 v4 beta here, I think there's not really a specific reason for picking 2 minutes, the only requirement is that `exp` must be a value no more than 5 minutes after the `iat`

Reply:
I don't think we currently support configuring this on any other SDK and I haven't seen any request to do so yet, so I'd maybe defer to adding if we get a request